core/local/layers/linux.js

```diff
@@ -0,0 +1,180 @@
+      const batch /*: LayerEvent[] */ = []
+      for (const entry of await fse.readdir(fullPath)) {
+        try {
+          const abspath = path.join(this.syncPath, relativePath, entry)
+          batch.push(await this.buildAddEvent(abspath))
+        } catch (err) {
+          // TODO error handling
+        }
+      }
+      // TODO ignore
+      if (batch.length === 0) {
+        return
+      }
+      this.next.process(batch)
+      for (const event of batch) {
+        if (event.doc.docType === 'folder') {
+          await this.watch(event.doc.path)
+        }
+      }
+    } catch (err) {
+      // The directory may been removed since we wanted to watch it
+    }
+  }
+
+  async process (events /*: AtomWatcherEvent[] */) {
+    // TODO logger
+    // TODO ignore
+    const batch /*: LayerEvent[] */ = []
+    for (const event of events) {
+      switch (event.action) {
+        case 'created':
+          const eAdd = await this.buildAddEvent(event.path)
+          batch.push(eAdd)
+          if (eAdd.doc.docType === 'folder') {
+            this.watch(eAdd.doc.path)
+          }
+          break
+        case 'modified':
+          batch.push(await this.buildUpdateEvent(event.path))
+          break
+        case 'deleted':
+          const eDel = await this.buildRemoveEvent(event.path, event.kind)
+          batch.push(eDel)
+          const w = this.watchers.get(eDel.doc.path)
+          if (w) {
+            w.dispose()
+            this.watchers.delete(eDel.doc.path)
+          }
+          break
+        case 'renamed':
+          batch.push(await this.buildMoveEvent(event.path, event.oldPath))
+          break
+        default:
+          throw new Error(`Unknown atom/watcher action ${event.action}`)
+      }
+    }
+    this.next.process(batch)
+  }
+
+  async buildAddEvent (abspath /*: string */) /*: Promise<LayerAddEvent> */ {
+    let doc /*: ?Metadata */
+    const fpath = this.relativePath(abspath)
+    const stats = await fse.stat(abspath)
+    if (stats != null && stats.isDirectory()) {
+      doc = buildDir(fpath, stats)
+    } else {
+      doc = buildFile(fpath, stats, '')
+    }
+    return { action: 'add', abspath, doc }
+  }
+
+  async buildUpdateEvent (abspath /*: string */) /*: Promise<LayerUpdateEvent> */ {
+    let doc /*: ?Metadata */
+    const fpath = this.relativePath(abspath)
+    const stats = await fse.stat(abspath)
+    if (stats != null && stats.isDirectory()) {
+      doc = buildDir(fpath, stats)
+    } else {
+      doc = buildFile(fpath, stats, '')
+    }
+    return { action: 'update', abspath, doc }
+  }
+
+  async buildRemoveEvent (abspath /*: string */, kind /*: string */) /*: Promise<LayerRemoveEvent> */ {
+    const fpath = this.relativePath(abspath)
+    let doc /*: ?Metadata */
+    if (kind === 'directory') {
+      doc = buildDir(fpath, new fs.Stats())
+    } else {
+      doc = buildFile(fpath, new fs.Stats(), '')
+    }
+    return { action: 'remove', abspath, doc }
```

Comment:
Previously we had `action: 'deleted'`, now we have also `action: 'remove'`... Isn't it a bit error-prone?

Reply:
I mean in the original atom-watcher event.

Reply:
Well, there is a mismatch between the naming from atom/watcher and what we use elsewhere (like names of the method of Prep/Merge). I've chosen to stick with the later, but if you prefer the former, I don't really mind.

Reply:
Well, I still find that atom/watcher `renamed` that can mean `moved` is not terrible. But if you prefer atom/watcher naming, I can live with that.

Reply:
I believe Prep/Merge use `deleteFile` / `deleteFolder`, not *remove* ;)